src/main/java/uk/gov/companieshouse/itemhandler/kafka/OrdersKafkaConsumer.java

```diff
@@ -1,173 +1,271 @@
+        Map<String, String> dataMap = new HashMap<>();
+        MessageHeaders messageHeaders = message.getHeaders();
+        dataMap.put("data.key", "" + messageHeaders.get(KafkaHeaders.RECEIVED_MESSAGE_KEY));
+        dataMap.put("data.topic", "" + messageHeaders.get(KafkaHeaders.RECEIVED_TOPIC));
+        dataMap.put("data.offset", "" + messageHeaders.get(KafkaHeaders.OFFSET));
+        dataMap.put("data.partition", "" + messageHeaders.get(KafkaHeaders.RECEIVED_PARTITION_ID));
+
+        return dataMap;
+    }
+
+    protected void republishMessageToTopic(String orderUri, String currentTopic, String nextTopic) {
+        LOGGER.info(String.format("Republishing message: \"%1$s\" received from topic: \"%2$s\" to topic: \"%3$s\"",
+                orderUri, currentTopic, nextTopic));
+        try {
+            kafkaProducer.sendMessage(createRetryMessage(orderUri, nextTopic));
+        } catch (ExecutionException | InterruptedException e) {
+            LOGGER.error(String.format("Error sending message: \"%1$s\" to topic: \"%2$s\". \"%n%3$s\"",
+                    orderUri, nextTopic, Arrays.toString(e.getStackTrace())));
+            if (e instanceof InterruptedException) {
+                Thread.currentThread().interrupt();
             }
         }
     }
 
-    protected Message createRetryMessage(String orderUri) throws SerializationException {
+    protected Message createRetryMessage(String orderUri, String topic) {
         final Message message = new Message();
         AvroSerializer serializer = serializerFactory.getGenericRecordSerializer(OrderReceived.class);
         OrderReceived orderReceived = new OrderReceived();
-        orderReceived.setOrderUri(orderUri);
+        orderReceived.setOrderUri(orderUri.trim());
 
-        message.setValue(serializer.toBinary(orderReceived));
-        message.setTopic(ORDER_RECEIVED_TOPIC_RETRY);
+        message.setKey(ORDER_RECEIVED_KEY_RETRY);
+        try {
+            message.setValue(serializer.toBinary(orderReceived));
+        } catch (SerializationException e) {
+            LOGGER.error(String.format("Error serializing message: \"%1$s\" for topic: \"%2$s\". \"%n%3$s\"",
+                    orderUri, topic, Arrays.toString(e.getStackTrace())));
+        }
+        message.setTopic(topic);
         message.setTimestamp(new Date().getTime());
 
         return message;
     }
 
-    private void logMessageReceived(String message, String topic){
-        LOGGER.info(String.format("Message: %1$s received on topic: %2$s", message, topic));
+    private static void updateErrorRecoveryOffset(long offset){
+        errorRecoveryOffset = offset;
+    }
+
+    /**
+     * Sets `errorRecoveryOffset` to latest topic offset (error topic) minus 1, before error consumer starts. This
+     * helps the error consumer to stop consuming messages when all messages up to `errorRecoveryOffset` are processed.
```

Comment:
`This stops the error consumer from consuming further messages...`?

Reply:
This method helps set the `errorRecoveryOffset` when the app is launched in error mode, just before the `-error` consumer starts consuming. So it helps the `-error` consumer stop consuming messages when its `offset` matches `errorRecoveryOffset`.

Reply:
I think it was really a question about how this is described. If you think this is more helping it to stop than ensuring it will stop when it has reached the right point, that's fine.